### ISSUE TITLE
`_arguments:comparguments:409: not enough arguments` on the integrated terminal

### ISSUE THREAD

**Issue by himkt** [microsoft/vscode#161159]
Does this issue occur when all extensions are disabled?: **Yes**

- VS Code Version: 1.71.2 (Commit: 74b1f979648cc44d385a2286793c226e611f59e7)
- OS Version: macOS Monterey 12.5

### Summary

The error `_arguments:comparguments:409: not enough arguments` shows on the `<TAB>` insertion in the integrated terminal of VSCode.

### Disclaimer

I'm not sure if here is a appropriate place but I want professionals to help for investigation...
**This problem only happens on VSCode. It doesn't replicate on some Apps such as Terminal.app or iTerm2.app.**
Other user also experienced this problem ([link](https://www.reddit.com/r/zsh/comments/vtmgn2/zsh_tab_completion_of_filenames_doesnt_work_well/)), he also said the problem only occurs on the integrated terminal on VSCode.

And I'm so sorry not to create a minimum reproducible configuration... I should had looked inside the zplug inside but I couldn't afford to the such further investigation. :bow:

### Steps to Reproduce:

1. Install zplug

```sh
mkdir -p $HOME/.config  # create XDG_CONFIG_HOME if it doesn't exist
git clone https://github.com/zplug/zplug $HOME/.config/zplug  # install zplug
```

2. configure zsh

Write the zsh configuration as follows:

`$HOME/.zshrc`

```sh
export ZPLUG_HOME=$HOME/.config/zplug
source $ZPLUG_HOME/init.zsh
```

3. Open integrated terminal

Run `Terminal: Create New Terminal` on the command palette on VSCode. Put `ls ` and press `<TAB>` to replicate the issue.

<img width="440" alt="Screen Shot 2022-09-17 at 10 54 23" src="https://user-images.githubusercontent.com/5164000/190835995-30880677-74f3-4259-b7de-9d234ea39808.png">

Note that this problem doesn't happen when no whitespace is given before `<TAB>`:

<img width="663" alt="Screen Shot 2022-09-17 at 10 54 16" src="https://user-images.githubusercontent.com/5164000/190835975-326c66c0-02d9-403b-a5fe-b49662fbfb76.png">

**Comment by himkt** [comment]
I've succeeded to replicate with the minimum config (simply running `compinit`).

`~/.zshrc`

```sh
autoload -U compinit
compinit
```

Then launching vscode and open the integrated terminal, it happens.
My Zsh is `5.8.1 (x86_64-apple-darwin21.0)`.

**Comment by himkt** [comment]
I found the cause. The active zsh was **v5.9** installed in homebrew dir (`/opt/homebrew/bin`), but `_arguments` in `/usr/share/zsh/5.8.1/functions` was loaded. After trying unlinking zsh on homebrew, the error disspeared.

I'm still not sure why the problem only happens in the integrated terminal of VSCode, not in iTerm2... 🤔

**Comment by solzard** [comment]
Hi @himkt, thank you for suggesting a workaround! 
After reading your idea, I configured to use preinstalled zsh on VSCode's internal terminal but keep using homebrew zsh on iTerm2.
`settings.json` below works fine for me.
```json
	"terminal.integrated.defaultProfile.osx": "preinstalled-zsh",
    "terminal.integrated.profiles.osx": {
        "preinstalled-zsh": {
            "path": "/bin/zsh", // Explicitly specify preinstalled zsh
        },
    },
```

The left is a previous homebrew-installed zsh, and the right is `/bin/zsh` called with the above settings.
![Screenshot 2022-09-19 at 10 23 23](https://user-images.githubusercontent.com/29563481/190936991-ee71cd78-f837-4867-b44e-3122dcf09d45.jpg)

---

By the way, how did you find `_argument` is loaded from `/usr/share/zsh/5.8.1/functions`?
`which _argument` only returns its function definition. 
I couldn't find where it's loaded from when investigated it months ago.

**Comment by himkt** [comment]
> settings.json below works fine for me.

Thank you! I will use this workaround as well.

---

```
> type _arguments
_arguments is an autoload shell function
> _arguments
> type _arguments
_arguments is a shell function from /usr/share/zsh/5.8.1/functions/_arguments
```

I used `type` command to find that. Once calling `_arguments`, `type` shows the location of the script where the function is defined.

**Comment by PhotonQuantum** [comment]
I believe this is related to `terminal.integrated.inheritEnv`. After disabling this config, `$FPATH` is correct now.

And this problem is related to
`export FPATH="/opt/homebrew/share/zsh/site-functions${FPATH+:$FPATH}";`
in my `.zshrc` file.

With this line removed, the `FPATH` is correct even with `inheritEnv` enabled.

**Comment by solzard** [comment]
@PhotonQuantum Can you show me your full `$FPATH` for reference?

**Comment by PhotonQuantum** [comment]
@solzard Sure.

I'm using Nix to manage my system, so the FPATH is a bit messy. I'm pasting it to gist.

https://gist.github.com/PhotonQuantum/ec6f5bffc531f5e17ea05299733fc6c9

**Comment by PhotonQuantum** [comment]
I believe I've figured out the cause.

tl;dr

Check `$SHELL` and `which zsh`. You may run `chsh -s $(which zsh)` to change your default shell if they don't match. If you don't want to change the default shell, read the workaround given by @solzard instead.

---

This is caused by the difference between your default login shell and the shell executable path in your `$PATH`.

VSCode uses a shell script to load environment variables during start, which is evaluated using the user login shell. This environment is inherited when the integrated terminal is loaded (if you have `inheritEnv` enabled). 

However, this terminal uses the shell executable in your `PATH`. If their version doesn't match, then you are in trouble.

**Comment by kytta** [comment]
> You may run `chsh -s $(which zsh)` to change your default shell if they don't match.

FYI: This may not work for some macOS users. For example, my machine has told me that the shell is non-standard. The solution is to:

1. Either do the command with sudo and your username at the end: `sudo chsh -s $(which zsh) $(whoami)`; or
2. Append the path to the desired Zsh executable to `/etc/shells` (not recommended, as it will confuse macOS at times and requires sudo any way)

**Comment by PhotonQuantum** [comment]
@kytta You are right.
A shell must be in /etc/shells to be set as login shell by chsh.
My /etc/shells is managed by nix and I forgot to mention this.
